Gb/bc

grantbuster opened the pull request:
Implemented more bias correction methods motivated by Sup3rCC. 

Example:
![map_bespoke_bc](https://github.com/NREL/rex/assets/12700356/f7edf4e8-9be8-4dd9-9765-4fc4b2d49216)

ppinchuk commented on rex/utilities/bc_utils.py:
```diff
@@ -0,0 +1,105 @@
+logger = logging.getLogger(__name__)
+
+
+def _parse_bc_table(bc_df, gids):
```
Is this intentionally private? In the docs, `bc_utils` will look like an empty module

grantbuster replied:
Good point. Its a tricky one because it causes a recursive import. I moved it to a new .py file, made it public, and added a real bc_utils.py file with actual utilities that can be imported by bias correction

ppinchuk commented on rex/utilities/bc_utils.py:
```diff
@@ -0,0 +1,105 @@
+        msg = ('Bias correction table provided, but "method" column not '
+               'found! Only see columns: {}. Need to specify "method" which '
+               'is a function name from `rex.bias_correction`'
+               .format(list(bc_df.columns)))
```
Did you want to add a `logger.error(msg)` call here like for the exception below?

grantbuster replied:
missed this, thanks!

ppinchuk commented on rex/utilities/bc_utils.py:
```diff
@@ -0,0 +1,105 @@
+    if not bool_bc.any():
+        return None, {}, bool_bc
+
+    elif not bool_bc.all():
```
`elif` can just be `if` here I think

grantbuster replied:
Kind of the same right? I thought elif might lower the pylint perceived function complexity lol but i switched it back to if

ppinchuk commented on rex/bias_correction.py:
```diff
@@ -0,0 +1,578 @@
+        """If dist='empirical', this will get the quantile values for the CDF
+        x-values specified in the input params"""
+
+        if self.sampling == 'linear':
```
This looks similar to code in the `cdf` function (from `temporal_stats.py`). Could consider pulling it out into it's own small utility function and calling that in both places?

grantbuster replied:
moved to the new actual bc_utils.py file

ppinchuk commented on rex/bias_correction.py:
```diff
@@ -0,0 +1,578 @@
+
+    def cdf(self, x, params):
+        """Run the CDF function e.g., convert physical variable to quantile"""
+
```
If you're at all interested in decreasing indentation, you could do 

```
if self.scipy_dist is not None:
    return self.scipy_dist.cdf(x, *params)

p = np.zeros_like(x)
for idx in range(x.shape[1]):
    xp = params[idx, :]
    fp = self._get_quantiles(len(xp))
    p[:, idx] = np.interp(x[:, idx], xp, fp)

return p
```

(same comment for `ppf` below)

grantbuster replied:
I didn't do this just because i kind of like assigning p/x to the scipy output so it's clear its equivalent to the empirical function.

ppinchuk reviewed (dismissed):
Man, that Quantile Delta Mapping bias correction looks SO GOOD!

Excellent job with the code and especially with the docstrings!!
I have added a few suggestions below, all of which can be ignored.

I didn't check any of the actual algorithm logic, since I'm not familiar with it. Maybe good to have Brandon take a look too if he know more about it?

grantbuster commented on rex/temporal_stats/temporal_stats.py:
```diff
@@ -83,6 +83,73 @@ def circular_mean(data, weights=None, degrees=True, axis=0,
+    nan_mask = np.isnan(data)
+    if nan_mask.all():
+        return np.zeros(n_samples)
+
```
done!

grantbuster commented:
> Man, that Quantile Delta Mapping bias correction looks SO GOOD!
> 
> Excellent job with the code and especially with the docstrings!! I have added a few suggestions below, all of which can be ignored.
> 
> I didn't check any of the actual algorithm logic, since I'm not familiar with it. Maybe good to have Brandon take a look too if he know more about it?

Thanks Paul! Generally i think this could be really useful for us in using ERA or HRRR as resource inputs for 2015-present. 

I shuffled some things around a little bit in response to your comments. The table parsing function is awkward because it creates a circular import so i moved it to its own file and created a real bc utils file. Added @bnb32  as he said he wants to take a look at the QDM method.

bnb32 commented on rex/utilities/bc_utils.py:
```diff
@@ -0,0 +1,294 @@
+        params_mh = self._clean_params(self.params_mh, arr.shape)
+        params_mf = self._clean_params(self.params_mf, arr.shape)
+
+        q_mf = self.cdf(arr, params_mf)  # Tau_m_p
```
if these comment refer to equation terms could you add the full equation in comments as well?

grantbuster replied:
Yeah gotcha. Will do but i'll also add the equation numbers. I'm not sure the equation terms are any more readable than the code. For example:

<img width="303" alt="image" src="https://github.com/NREL/rex/assets/12700356/afceea56-7e22-458d-ba09-0481cf82d3d5">

vs. 

`x_oh = self.ppf(q_mf, params_oh)  # x^_o:m_h:p`

bnb32 replied:
But is this a term in a larger equation or no? I was thinking they were terms and that a complete equation would provide context but maybe thats not true.

grantbuster replied:
There's only one code line that is just a part of a larger equation from cannon. I added more comments, see the new commit. I think it's as good as we can get.

bnb32 reviewed (approved):
This is suprisingly simple. Nice implementation. Just a single minor comment so appoving prior to that one change.

bnb32 reviewed (dismissed):
This is suprisingly simple. Nice implementation. Just a single minor comment so appoving prior to that one change.
